app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.test.ts

```diff
@@ -0,0 +1,756 @@
+            value: {
+              requiredScopes: {
+                'eip155:1': {
+                  accounts: ['eip155:1:0x1', 'eip155:1:0x2'],
+                },
+                'eip155:5': {
+                  accounts: ['eip155:5:0x1', 'eip155:5:0x3'],
+                },
+              },
+              optionalScopes: {
+                'eip155:1': {
+                  accounts: ['eip155:1:0x4'],
+                },
+                'other:1': {
+                  accounts: ['other:1:0x4'],
+                },
+              },
+            },
+            isMultichainOrigin: false,
+          },
+        ],
+      },
+    }),
+  );
+  const updateCaveat = jest.fn();
+  const grantPermissions = jest.fn().mockReturnValue(
+    Object.freeze({
+      [Caip25EndowmentPermissionName]: {
+        id: 'new',
+        parentCapability: Caip25EndowmentPermissionName,
+        caveats: [
+          {
+            type: Caip25CaveatType,
+            value: {
+              requiredScopes: {},
+              optionalScopes: {},
+            },
+          },
+        ],
+      },
+    }),
+  );
+  const requestPermissionApprovalForOrigin = jest.fn().mockResolvedValue({
+    approvedChainIds: ['0x1', '0x5'],
+    approvedAccounts: ['0xdeadbeef'],
+  });
+  const getAccounts = jest.fn().mockReturnValue([]);
+  const response: PendingJsonRpcResponse<Json> = {
+    jsonrpc: '2.0' as const,
+    id: 0,
+  };
+  const handler = (request: unknown) =>
+    requestPermissionsHandler.implementation(
+      request as JsonRpcRequest<[RequestedPermissions]> & { origin: string },
+      response,
+      next,
+      end,
+      {
+        requestPermissionsForOrigin,
+        getPermissionsForOrigin,
+        updateCaveat,
+        grantPermissions,
+        requestPermissionApprovalForOrigin,
+        getAccounts,
+      },
+    );
+
+  return {
+    response,
+    next,
+    end,
+    requestPermissionsForOrigin,
+    getPermissionsForOrigin,
+    updateCaveat,
+    grantPermissions,
+    requestPermissionApprovalForOrigin,
+    getAccounts,
+    handler,
+  };
+};
+
+describe('requestPermissionsHandler', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  beforeEach(() => {
+    MockMultichain.setEthAccounts.mockImplementation(
+      (caveatValue) => caveatValue,
+    );
+    MockMultichain.setPermittedEthChainIds.mockImplementation(
+      (caveatValue) => caveatValue,
+    );
+  });
+
+  it('returns an error if params is malformed', async () => {
+    const { handler, end } = createMockedHandler();
+
+    const malformedRequest = {
+      ...getBaseRequest(),
+      params: [],
+    };
+    await handler(malformedRequest);
+    expect(end).toHaveBeenCalledWith(
+      invalidParams({ data: { request: malformedRequest } }),
+    );
+  });
+
+  it('requests approval from the ApprovalController for eth_accounts and permittedChains when only eth_accounts is specified in params and origin is not snapId', async () => {
```

Comment:
I'm not seeing that `origin` is not snapId in this test, thoughts on making that more clear?

Reply:
https://github.com/MetaMask/metamask-extension/pull/27847/commits/f7a09fda54e6e4998b098b2ea84e8557b0e13d3d